### DISCUSSION TITLE
String matching not working

### DISCUSSION THREAD

**Discussion by ilmari-lauhakangas** [textX/textX#370] [Q&A] [answered]
We use textX in LibreOffice for creating UI tests out of logged actions. I ran into an issue while trying to adjust a grammar file.

This is the file: https://git.libreoffice.org/core/+/refs/heads/master/uitest/ui_logger_dsl/UI_Object_commands.tx

It has TabControlUIObject with tab_page_number=INT. It turns out some of our tab control items have a string as ID instead of a number.  I changed the type to STRING and ran into a matching problem.

This is the log file that is being parsed, **Choose Tab number internet in 'tabcontrol' from** is the relevant line:

```
Start writer
Send UNO Command (".uno:UpdateInputFields") 
Type on writer {"KEYCODE": "CTRL+k"}
Send UNO Command (".uno:HyperlinkDialog") 
Choose Tab number internet in 'tabcontrol' from
Open Modeless HyperlinkDialog
Click on 'cancel' from HyperlinkDialog
Click on 'cancel' from HyperlinkDialog
Close Dialog
Type on writer {"KEYCODE": "CTRL+q"}
Send UNO Command (".uno:Quit") 
```

In the [dsl_core.py](https://git.libreoffice.org/core/+/refs/heads/master/uitest/ui_logger_dsl/dsl_core.py), I turned debugging on with

`self.ui_dsl_mm = metamodel_from_file("ui_logger_dsl_grammar.tx", debug=True)`

In the debug output, I see this:

```
>> Matching rule TabControlUIObject=Sequence in UIObjectCommand at position 136 => nkDialog")*  Choose T
    ?? Try match rule StrMatch(Choose Tab number) in TabControlUIObject at position 138 => Dialog")  *Choose Tab
    ++ Match 'Choose Tab number' at 138 => 'Dialog")  *Choose Tab number*'
    >> Matching rule __asgn_plain=Sequence[tab_page_number] in TabControlUIObject at position 155 => Tab number* internet 
        ?? Try match rule STRING=RegExMatch(("(\\"|[^"])*")|(\'(\\\'|[^\'])*\')) in __asgn_plain at position 156 => ab number *internet i
        ?? Try match rule Comment=RegExMatch(\/\/.*$) in __asgn_plain at position 156 => ab number *internet i
        -- NoMatch at 156
        -- NoMatch at 156
    <<- Not matched rule __asgn_plain=Sequence[tab_page_number] in __asgn_plain at position 155 => Tab number* internet 
<<- Not matched rule TabControlUIObject=Sequence in TabControlUIObject at position 136 => nkDialog")*  Choose T
```

What gives? Why is it not grabbing 'internet' as the string?

I am using textX 2.3.0 on Arch Linux.

**Comment by igordejanovic** [comment, marked as answer]
Built-in `STRING` rule [matches quoted strings](https://github.com/textX/textX/blob/master/textx/lang.py#L116). `internet` would be matched by the `ID` rule. But number won't be matched by `ID`. So, probably the best would be to make a new rule that matches either `ID` or `INT` and use that for `tab_page_number`.

**Reply by ilmari-lauhakangas** [reply]
Thanks a lot, I will look into it.

**Reply by ilmari-lauhakangas** [reply]
Yay, it worked :) I added an [abstract rule](https://textx.github.io/textX/stable/grammar/#rule-types) to the .tx file:

```
tab_page_number:
  INT | ID
;
```

and then just dropped the type assignment from the existing line.

**Reply by igordejanovic** [reply]
Just don't forget, if you wan't to keep what is matched on the object you have to do `tab_page_number=tab_page_number`. Left side is the name of the attribute that will keep the matched value while the right side if the name of the rule. BTW, to be consistent with the rest of the grammar you should probably name the rule `TabPageNumber`.
